### PR TITLE
fix equal checks for cells and output

### PR DESCRIPTION
**Pull request by sokra**
### Description

* Check for equal Vc on returning a value
* Avoid PartialEq checks for values that are never equal

### CODE REVIEW

File: crates/turbopack-core/src/source_map/mod.rs [hunk 1]
```diff
@@ -43,7 +43,7 @@ pub trait GenerateSourceMap {
 /// The distinction between the source map spec's [sourcemap::Index] and our
 /// [SourceMap::Sectioned] is whether the sections are represented with Vcs
 /// pointers.
-#[turbo_tasks::value(shared)]
+#[turbo_tasks::value(shared, cell = "new")]
```
Comment (sokra) [inline crates/turbopack-core/src/source_map/mod.rs:46]: `cell = "new"` changes the behavior of `.cell()` to no longer perform an `==` check and just always override the cell.

File: crates/turbo-tasks-memory/src/output.rs [hunk 2]
```diff
@@ -56,7 +56,17 @@ impl Output {
 
     pub fn link(&mut self, target: RawVc, turbo_tasks: &dyn TurboTasksBackendApi<MemoryBackend>) {
         debug_assert!(*self != target);
-        self.assign(OutputContent::Link(target), turbo_tasks)
+        if let OutputContent::Link(old_target) = &self.content {
+            if *old_target == target {
+                // unchanged
+                return;
```
Comment (sokra) [inline crates/turbo-tasks-memory/src/output.rs:62]: We can early return when returning the same value again.

Previously this notified all dependent tasks, which is unnecessary.

File: crates/turbopack-core/src/source_map/mod.rs (hunk 1 above)
Comment (arlyon) [reply crates/turbopack-core/src/source_map/mod.rs:46]: How do we determine which cells should always be overridden?

Comment (bgw) [reply crates/turbopack-core/src/source_map/mod.rs:46]: When you don't want cells to ever match using equality? Which I imagine should be never unless you've got another bug (e.g. inconsistent cell creation order?).

I agree this seems weird and like we're masking bigger problems.

**Review by bgw — APPROVED** [review]

File: crates/turbopack-core/src/source_map/mod.rs (hunk 1 above)
Comment (sokra) [reply crates/turbopack-core/src/source_map/mod.rs:46]: > When you don't want cells to ever match using equality?

* some types have every expensive equal checks and are very unlikely to be equal (e. g. the parsed AST or here the SourceMap)
* some types don't have a equality defined
* some types define equality in a way that it never matches

But in general you want to check equality for early stop bubbling invaliations...

**Pull request merged**